### ISSUE TITLE
Scan commits instead of file tree for todos

### ISSUE THREAD

**Issue by Husterknupp** [Husterknupp/jubilant-todo-tribble#16]
.. rather than scanning the whole file tree over and over again

- GitHub commit model (see `patch`) https://api.github.com/repos/Husterknupp/jubilant-todo-tribble/commits/30fd93abf61d32ae9132def3f01c67aa76dc4da2
- Gitlab commit model https://docs.gitlab.com/ce/api/commits.html#get-a-single-commit
- Jgit Java library for instance https://github.com/eclipse/jgit/blob/master/org.eclipse.jgit/src/org/eclipse/jgit/patch/Patch.java#L136

**Comment by Husterknupp** [comment]
- makes sense because for bigger projects a whole file tree scanning is way more expensive than checking for new commits
- might be complicated though b/c git diff entities look complicated on first glance. Need to investigate

**Comment by Husterknupp** [comment]
For new repos
1. remember current timestamp in format `2012-09-20T11:50:22+03:00`
2. scan and save whole file tree for todos as usual
3. mark for this repo somewhere 'initialScanDone: true'
4. save remembered timestamp along ('lastIntervalScan')

For all known repos (initialScanDone: true)
4.  remember current timestamp
5.  read remembered timestamp (lastIntervalScan)
6.  read commits since then, scan and update todos in diffs
7.  save remembered timestamp in lastIntervalScan

**Comment by Husterknupp** [comment]
How to interpret the git *diff hunks*
* does a given diff hunk mean, a todo was removed?
  * if yes, what line was it before it was removed?
* if it was added, what line is the new todo?

**Comment by Husterknupp** [comment]
after evaluating 
 * `com.googlecode.java-diff-utils` (not listing the added/removed lines in a convenient way)
 * and `com.zutubi.diff` (not finding more than one hunk per file)

I'll go with Jgit.

Before implementation, I should start by writing a unit test. Verify a todos can be added and removed in a diff with new and removed lines respectively

**Comment by Husterknupp** [comment]
@Bantolomeus we will have another yaml property. `gitlab.repoBranch` will make todo finding more accurate. Please add it to your local yaml once this branch is finished

**Comment by Husterknupp** [comment]
I will have to save 
- a successful scan date to the file system as well as 
- restructure the todo history to feature todos grouped by repoId + branchName

**Comment by Bantolomeus** [comment]
@Husterknupp Do you have problems? Can I solve them for you?

**Comment by Husterknupp** [comment]
@Bantolomeus thanks buddy. I listed two separate tasks above. If you find time, go ahead and pick one

**Comment by Husterknupp** [comment]
"Scan commits instead of file tree for todos"

that's what the title says and that's what the app is doing. Still new todos are not being pushed as new entry in Jira, but that's another story. Scope of the ticket is reached
